### ISSUE TITLE
Multi-entity save can cause unexpected effects with save_post action hooks

### ISSUE THREAD

**Issue by carlomanf** [WordPress/gutenberg#29700]
## Description
When using the "template mode" of the post editor, and a reusable block is inside the template, and the reusable block is updated, the `save_post_wp_block` action hook doesn't appear to fire.

## Step-by-step reproduction instructions

1. Put a reusable block inside a `singular` template
2. Open the post editor for any post that uses that template
3. Open the "template mode"
4. Edit the reusable block
5. Save changes
6. Observe that the `save_post_wp_block` action hook does not appear to fire

## Expected behaviour
The `save_post_wp_block` action hook should fire, since the post was saved.

## Actual behaviour
The callback I was using on the `save_post_wp_block` action hook didn't take effect, when the reusable block was updated through the template mode. I am guessing this was because the hook didn't fire, but I didn't have time to verify this. When updating the reusable block through the site editor or the normal post editor, my callback completed properly, so I can verify it affects the "template mode" editor only.

**Comment by carlomanf** [comment]
I had some time to test this more thoroughly, and I determined that it was caused by the order of precedence of entities being saved in the multi-entity save. In other words, my callback for the reusable block was expecting the main post to NOT be saved yet, but actually the main post had already been saved before the reusable block got saved.

I'm not sure anything needs to be done to "fix" this, but it would be helpful for some documentation for plugin developers, about the order in which the entities (main post, template, template part, reusable block etc) get saved in a multi-entity save. Also, it'd be very important to keep the order of precedence consistent. For example, you wouldn't want the reusable block saved before the main post in one editor, and the main post saved before the reusable block in another editor.

**Comment by tyxla** [comment]
I agree it would be useful to have this better documented 👍 Keeping open for now.
